include/nbl/builtin/hlsl/emulated/float64_t.hlsl

```diff
@@ -0,0 +1,657 @@
+#ifndef _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_HLSL_INCLUDED_
+#define _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_HLSL_INCLUDED_
+
+#include <nbl/builtin/hlsl/emulated/float64_t_impl.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+    /*enum E_ROUNDING_MODE
+    {
+        FLOAT_ROUND_NEAREST_EVEN,
+        FLOAT_ROUND_TO_ZERO,
+        FLOAT_ROUND_DOWN,
+        FLOAT_ROUND_UP
+    };*/
+
+    // currently only FLOAT_ROUND_TO_ZERO is supported, cannot implement partial specialization in this case due to dxc bug https://github.com/microsoft/DirectXShaderCompiler/issues/5563
+    // TODO: partial specializations with new template parameter `E_ROUNDING_MODE RoundingMode`
+    template<bool FastMath = true, bool FlushDenormToZero = true>
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+        using this_t = emulated_float64_t<FastMath, FlushDenormToZero>;
+
+        storage_t data;
+
+        // constructors
+        /*static emulated_float64_t create(uint16_t val)
+        {
+            return emulated_float64_t(bit_cast<uint64_t>(float64_t(val)));
+        }*/
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(this_t val)
+        {
+            return val;
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(int32_t val)
+        {
+            return bit_cast<this_t>(emulated_float64_t_impl::castToUint64WithFloat64BitPattern(int64_t(val)));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(int64_t val)
+        {
+            return bit_cast<this_t>(emulated_float64_t_impl::castToUint64WithFloat64BitPattern(val));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(uint32_t val)
+        {
+            return bit_cast<this_t>(emulated_float64_t_impl::castToUint64WithFloat64BitPattern(uint64_t(val)));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(uint64_t val)
+        {
+            return bit_cast<this_t>(emulated_float64_t_impl::castToUint64WithFloat64BitPattern(val));
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(float32_t val)
+        {
+            this_t output;
+            output.data = emulated_float64_t_impl::castFloat32ToStorageType<FlushDenormToZero>(val);
+            return output;
+        }
+
+        NBL_CONSTEXPR_STATIC_INLINE this_t create(float64_t val)
```

Comment:
why have both `create` and specializations of `static_cast`?

Reply:
why not? also this is how `_static_cast` from `T` to `emulated_float64_t<FastMath, FlushDenormToZero>` looks like:

```cpp
template<typename From, bool FastMath, bool FlushDenormToZero>
struct static_cast_helper<emulated_float64_t<FastMath, FlushDenormToZero>, From, void>
{
    using To = emulated_float64_t<FastMath, FlushDenormToZero>;

    static inline To cast(From v)
    {
        return To::create(v);
    }
};
```

Reply:
ok, we couldn't specialize `_static_cast` for `emulated_float64_t` before the `emulated_float64_t` class is defined, which would require all methods calling `create` to be defined outside the class cause they would need to call `static_cast` instead after `emulated_float64_t` is defined and `static_cast_helper` is partially specialized

The only failure point is that `create(no_overload_for_this_t)` will compile as `create(some_implicit_conversion)` without warnings, e.g. if you forgot `int16_t` overload, it would go through some other overload.

Whereas `static_cast` will never do that due to template param matching.